### DISCUSSION TITLE
Any OData connector available?

### DISCUSSION THREAD

**Discussion by JessHolle** [trinodb/trino#20630] [Ideas]
It seems like OData provides a general purpose (limited) SQL-like REST interface behind which many disparate systems could hide.

I note that Teiid provided an OData translator for this purpose (https://teiid.github.io/teiid-documents/15.0.x/content/reference/r_odata-translator.html) separate from its OpenAPI translator (https://teiid.github.io/teiid-documents/15.0.x/content/reference/r_openapi-translator.html).  Presumably this is because the nature of OData provides a better framework for criteria /sorting push down, etc.

Am I missing something or is there no such connector for Trino?

**Comment by nineinchnick** [comment]
I've created an OpenAPI connector: https://github.com/nineinchnick/trino-openapi

I'd consider writing one for OData if I knew some cool APIs to use it with. Can you provide some examples?

**Reply by JessHolle** [reply]
The product I work on (Windchill) has a REST OData API.  Salesforce does as well as I recall, though perhaps there is some other API being pursued on that front.

What seems compelling about OData overall in this context is the fact that OData already supports a limited subset of SQL-like query capabilities in a standard fashion, thus providing a uniform basis for push-down, etc.  Of course, if a particular OData source don't actually support all of this then there is still a need to communicate this gap to Trino for that source.

**Reply by nineinchnick** [reply]
Is the Windchill API publicly available?

**Reply by JessHolle** [reply]
Well, it would appear (from use of an Icognito window) that docs are available even to non-customers via https://support.ptc.com/help/windchill_rest_services/r2.7/en/.

That said, it would seem the API catalog is available via Swagger, but from an installation of the software, which isn't therefore publicly available.

(This isn't really my own area, so I frankly get a bit lost is CSDL vs. OpenAPI and so on here.  It just _seems_ to me that like something like Trino should be leveraging OData nature of service when it obeying OData...)

**Reply by durban77** [reply]
The coolest of publicly available OData services as far as I know:

- http://services.odata.org/V4/OData/OData.svc/
- https://services.odata.org/V3/Northwind/Northwind.svc/
- http://services.odata.org/TripPinRESTierService/(S(readwrite))/
